File: 3Sum/Solution.java
public class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer> > res = new ArrayList<List<Integer> >();
        int size = nums.length;
        if (size < 3) return res;
        Arrays.sort(nums);
        int prev = 0;
        int sum = 0;
        for (int i = 0; i < size - 2;) {
            for (int j = i + 1, k = size - 1; j < k; ) {
                sum = nums[i] + nums[j] + nums[k];
                if (sum == 0) {
                    List<Integer> r = new ArrayList<Integer>();
                    r.add(nums[i]);
                    r.add(nums[j]);
                    r.add(nums[k]);
                    res.add(r);
                    prev = nums[j];
                    while (j < k && nums[j] == prev) j++;
                } else if (sum < 0) {
                    prev = nums[j];
                    while (j < k && nums[j] == prev) j++;
                } else {
                    prev = nums[k];
                    while (j < k && nums[k] == prev) k--;
                }
            }
            prev = nums[i];
            while (i < size && nums[i] == prev) i++;
        }
        return res;
    }
}

File: 3Sum/Solution.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        l = len(nums)
        if l < 3:
            return res
        list.sort(nums)
        i = 0
        while i < l - 2:
            j = i + 1
            k = l - 1
            while j < k:
                sum = nums[i] + nums[j] + nums[k]
                if sum == 0:
                    r = []
                    r.append(nums[i])
                    r.append(nums[j])
                    r.append(nums[k])
                    res.append(r)
                    prev = nums[j]
                    while j < k and nums[j] == prev:
                        j += 1
                elif sum < 0:
                    prev = nums[j]
                    while j < k and nums[j] == prev:
                        j += 1
                else:
                    prev = nums[k]
                    while j < k and nums[k] == prev:
                        k -= 1
            prev = nums[i]
            while i < l and nums[i] == prev:
                i += 1
        return res
